sdks/node-ts/src/apache_beam/core.ts

```diff
@@ -1,39 +1,173 @@
+const runnerApi = require('./proto/beam_runner_api');
+const translations = require('./internal/translations')
+
+
+// TODO(pabloem): Use something better, hah.
+var _pcollection_counter = 0;
+
+/**
+ * Represents an 'edge' in a graph. These may be PCollections, PCollection views,
+ * and Pipelines themselves.
+ */
 class PValue {
-    constructor() {
+    // TODO: Have a reference to its graph representation
+    type: string = "unknown";
+    name: string;
+
+    constructor(name: string) {
+        this.name = name;
+    }
+
+    isPipeline(): boolean {
+        return this.type === "pipeline";
     }
 
     apply(transform: PTransform): PValue {
         return transform.expand(this);
     }
 
-    map(callable: any): PValue {
+    map(callable: DoFn | GenericCallable): PValue {
         return this.apply(new ParDo(callable));
     }
 
+    // Top-level functions:
+    // - flatMap
+    // - filter?
 }
 
-class Pipeline extends PValue {
-
+/**
+ * A Pipeline is the base object to start building a Beam DAG. It is the
+ * first object that a user creates, and then they may start applying
+ * transformations to it to build a DAG.
+ */
+export class Pipeline extends PValue {
+    type: string = "pipeline";
+    portablePipeline: any;
 }
 
-class PCollection extends PValue {
+export class PCollection extends PValue {
+    type: string = "pcollection";
+    portablePcollection: any;
 
+    constructor(name: string, portablePcollection: any) {
+        super(name)
+        this.portablePcollection = portablePcollection;
+    }
 }
 
-class PTransform {
+export class PTransform {
     expand(input: PValue): PValue {
         throw new Error('Method expand has not been implemented.');
     }
 }
 
-class ParDo extends PTransform {
+
+/**
+ * @returns true if the input is a `DoFn`.
+ * 
+ * Since type information is lost at runtime, we check the object's attributes
+ * to determine whether it's a DoFn or not.
+ * 
+ * @example
+ * Prints "true" for a new `DoFn` but "false" for a function:
+ * ```ts
+ * console.log(new DoFn());
+ * console.log(in => in * 2));
+ * ```
+ * @param callableOrDoFn 
+ * @returns 
+ */
+function isDoFn(callableOrDoFn: DoFn | GenericCallable) {
+    const df = (callableOrDoFn as DoFn)
+    if (df.type !== undefined && df.type === "dofn") {
+        return true;
```

Comment:
We can return the value rather than `if (condition) return true else return false`.